dipy/reconst/ivim.py

```diff
@@ -153,20 +166,18 @@ def nlls_fit(data, gtab, guess_params):
 
     Returns
     -------
-    nlls_params: the S0, f, D_star, D value for each voxel.
+    ivim_params: the S0, f, D_star, D value for each voxel.
 
     """
     flat_data = data.reshape((-1, data.shape[-1]))
     # Flatten for the iteration over voxels:
     bvals = gtab.bvals
     ivim_params = np.empty((flat_data.shape[0], 4))
     for vox in range(flat_data.shape[0]):
-        popt, pcov = curve_fit(ivim_function, bvals, flat_data[vox],
-                               guess_params[vox])
-        ivim_params[vox, :4] = popt
+        res = leastsq(_ivim_error,
```

Comment:
It looks pretty close, though. Would be good to investigate that one. Do
you know how to run the tests with a debugger?

    `nosetests --pdb`

When an error is raised, that would drop you into pdb
<https://docs.python.org/2/library/pdb.html#>, and you can work
interactively in there to figure out what's going on. In particular, I
think that the default optimization method here is BFGS, but you might want
to try other methods.

On Thu, Jun 2, 2016 at 8:02 AM, Shahnawaz Ahmed <notifications@github.com>
wrote:

> In dipy/reconst/ivim.py
> <https://github.com/nipy/dipy/pull/1058#discussion_r65556786>:
>
> >
> >      """
> >      flat_data = data.reshape((-1, data.shape[-1]))
> >      # Flatten for the iteration over voxels:
> >      bvals = gtab.bvals
> >      ivim_params = np.empty((flat_data.shape[0], 4))
> >      for vox in range(flat_data.shape[0]):
> > -        popt, pcov = curve_fit(ivim_function, bvals, flat_data[vox],
> > -                               guess_params[vox])
> > -        ivim_params[vox, :4] = popt
> > +        res = leastsq(_ivim_error,
>
> Running local tests gives poor results and a nan mis-match with minimize.
>
> Arrays are not almost equal to 6 decimals
>
> x and y nan location mismatch:
>  x: array([[  9.99961548e-01,   9.86742110e-01,   9.73894304e-01,
>           9.61397992e-01,   9.49234352e-01,   9.25835852e-01,
>           9.03571317e-01,   8.82329225e-01,   8.62011942e-01,...
>  y: array([[ 1.        ,  0.98675793,  0.97389302,  0.96138429,  0.94921219,
>          0.92580612,  0.90354296,  0.88230778,  0.86200033,  0.84253307,
>          0.82382952,  0.80582274,  0.7884541 ,  0.77167213,  0.75543166,...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/dipy/pull/1058/files/c955cab175ec32c335e7ccd641ef9c46973418fa..834c2da60b66d8daf228cd45cdddf60152af9079#r65556786>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHPNigbTNC72Xpu6UnvN_YWpyGPTU6zks5qHvBugaJpZM4Imo2R>
> .
>